tsnet: data race in TestFallbackTCPHandler

```
==================
WARNING: DATA RACE
Read at 0x00c009a10cc8 by goroutine 1748:
  tailscale.com/tsnet.TestFallbackTCPHandler()
      tailscale.com/tsnet/tsnet_test.go:664 +0x4d5
  testing.tRunner()
      testing/testing.go:1595 +0x238
  testing.(*T).Run.func1()
      testing/testing.go:1648 +0x44

Previous write at 0x00c009a10cc8 by goroutine [206](https://github.com/tailscale/tailscale/actions/runs/6520601856/job/17708291457?pr=9804#step:8:207)0:
  tailscale.com/tsnet.TestFallbackTCPHandler.func1()
      tailscale.com/tsnet/tsnet_test.go:657 +0x44
  tailscale.com/tsnet.(*Server).getTCPHandlerForFlow()
      tailscale.com/tsnet/tsnet.go:777 +0x307
  tailscale.com/tsnet.(*Server).getTCPHandlerForFlow-fm()
      <autogenerated>:1 +0xb6
  tailscale.com/wgengine/netstack.(*Impl).acceptTCP()
      tailscale.com/wgengine/netstack/netstack.go:921 +0x8d7
  tailscale.com/wgengine/netstack.(*Impl).acceptTCP-fm()
      <autogenerated>:1 +0x3d
  gvisor.dev/gvisor/pkg/tcpip/transport/tcp.(*Forwarder).HandlePacket.func3()
      gvisor.dev/gvisor@v0.0.0-20230928000133-4fe30062272c/pkg/tcpip/transport/tcp/forwarder.go:98 +0x41

Goroutine 1748 (running) created at:
  testing.(*T).Run()
      testing/testing.go:1648 +0x82a
  testing.runTests.func1()
      testing/testing.go:2054 +0x84
  testing.tRunner()
      testing/testing.go:1595 +0x238
  testing.runTests()
      testing/testing.go:2052 +0x896
  testing.(*M).Run()
      testing/testing.go:1925 +0xb57
  main.main()
      _testmain.go:67 +0x2bd

Goroutine 2060 (finished) created at:
  gvisor.dev/gvisor/pkg/tcpip/transport/tcp.(*Forwarder).HandlePacket()
      gvisor.dev/gvisor@v0.0.0-20230928000133-4fe30062272c/pkg/tcpip/transport/tcp/forwarder.go:98 +0x577
  gvisor.dev/gvisor/pkg/tcpip/transport/tcp.(*Forwarder).HandlePacket-fm()
      <autogenerated>:1 +0x6d
  tailscale.com/wgengine/netstack.(*Impl).Start.(*Impl).wrapProtoHandler.func1()
      tailscale.com/wgengine/netstack/netstack.go:258 +0x10c
  gvisor.dev/gvisor/pkg/tcpip/stack.(*nic).DeliverTransportPacket()
      gvisor.dev/gvisor@v0.0.0-20230928000133-4fe30062272c/pkg/tcpip/stack/nic.go:838 +0x5be
  gvisor.dev/gvisor/pkg/tcpip/network/ipv4.(*endpoint).deliverPacketLocally()
      gvisor.dev/gvisor@v0.0.0-20230928000133-4fe30062272c/pkg/tcpip/network/ipv4/ipv4.go:1323 +0x10c1
  gvisor.dev/gvisor/pkg/tcpip/network/ipv4.(*endpoint).handleValidatedPacket()
      gvisor.dev/gvisor@v0.0.0-20230928000133-4fe30062272c/pkg/tcpip/network/ipv4/ipv4.go:1154 +0xdfc
  gvisor.dev/gvisor/pkg/tcpip/network/ipv4.(*endpoint).HandlePacket()
      gvisor.dev/gvisor@v0.0.0-20230928000133-4fe30062272c/pkg/tcpip/network/ipv4/ipv4.go:860 +0x6a4
  gvisor.dev/gvisor/pkg/tcpip/stack.(*groDispatcher).dispatch()
      gvisor.dev/gvisor@v0.0.0-20230928000133-4fe30062272c/pkg/tcpip/stack/gro.go:416 +0xc8
  gvisor.dev/gvisor/pkg/tcpip/stack.(*nic).DeliverNetworkPacket()
      gvisor.dev/gvisor@v0.0.0-20230928000133-4fe30062272c/pkg/tcpip/stack/nic.go:738 +0x[207](https://github.com/tailscale/tailscale/actions/runs/6520601856/job/17708291457?pr=9804#step:8:208)
  gvisor.dev/gvisor/pkg/tcpip/link/channel.(*Endpoint).InjectInbound()
      gvisor.dev/gvisor@v0.0.0-20230928000133-4fe30062272c/pkg/tcpip/link/channel/channel.go:202 +0x98
  tailscale.com/wgengine/netstack.(*Impl).injectInbound()
      tailscale.com/wgengine/netstack/netstack.go:746 +0x284
  tailscale.com/wgengine/netstack.(*Impl).injectInbound-fm()
      <autogenerated>:1 +0x47
  tailscale.com/net/tstun.(*Wrapper).filterPacketInboundFromWireGuard()
      tailscale.com/net/tstun/wrap.go:1050 +0xd04
  tailscale.com/net/tstun.(*Wrapper).Write()
      tailscale.com/net/tstun/wrap.go:1070 +0x2e8
  github.com/tailscale/wireguard-go/device.(*Peer).RoutineSequentialReceiver()
      github.com/tailscale/wireguard-go@v0.0.0-20230929[223](https://github.com/tailscale/tailscale/actions/runs/6520601856/job/17708291457?pr=9804#step:8:224)258-2f6748dc88e7/device/receive.go:516 +0x357
  github.com/tailscale/wireguard-go/device.(*Peer).Start.func3()
      github.com/tailscale/wireguard-go@v0.0.0-20[230](https://github.com/tailscale/tailscale/actions/runs/6520601856/job/17708291457?pr=9804#step:8:231)9292[232](https://github.com/tailscale/tailscale/actions/runs/6520601856/job/17708291457?pr=9804#step:8:233)58-2f6748dc88e7/device/peer.go:199 +0x44
==================
    stuntest.go:63: STUN server shutdown
    testing.go:1465: race detected during execution of test
--- FAIL: TestFallbackTCPHandler (0.95s)
FAIL	tailscale.com/tsnet
```

--- a/tsnet/tsnet_test.go
+++ b/tsnet/tsnet_test.go
@@ -26,6 +26,7 @@ import (
 	"reflect"
 	"strings"
 	"sync"
+	"sync/atomic"
 	"testing"
 	"time"
 
@@ -652,23 +653,23 @@ func TestFallbackTCPHandler(t *testing.T) {
 	}
 	t.Logf("ping success: %#+v", res)
 
-	s1TcpConnCount := 0
+	var s1TcpConnCount atomic.Int32
 	deregister := s1.RegisterFallbackTCPHandler(func(src, dst netip.AddrPort) (handler func(net.Conn), intercept bool) {
-		s1TcpConnCount++
+		s1TcpConnCount.Add(1)
 		return nil, false
 	})
 
-	if _, err = s2.Dial(ctx, "tcp", fmt.Sprintf("%s:8081", s1ip)); err == nil {
+	if _, err := s2.Dial(ctx, "tcp", fmt.Sprintf("%s:8081", s1ip)); err == nil {
 		t.Fatal("Expected dial error because fallback handler did not intercept")
 	}
-	if s1TcpConnCount != 1 {
-		t.Errorf("s1TcpConnCount = %d, want %d", s1TcpConnCount, 1)
+	if got := s1TcpConnCount.Load(); got != 1 {
+		t.Errorf("s1TcpConnCount = %d, want %d", got, 1)
 	}
 	deregister()
-	if _, err = s2.Dial(ctx, "tcp", fmt.Sprintf("%s:8081", s1ip)); err == nil {
+	if _, err := s2.Dial(ctx, "tcp", fmt.Sprintf("%s:8081", s1ip)); err == nil {
 		t.Fatal("Expected dial error because nothing would intercept")
 	}
-	if s1TcpConnCount != 1 {
-		t.Errorf("s1TcpConnCount = %d, want %d", s1TcpConnCount, 1)
+	if got := s1TcpConnCount.Load(); got != 1 {
+		t.Errorf("s1TcpConnCount = %d, want %d", got, 1)
 	}
 }

